### ISSUE TITLE
Figure out full text indices for localized text tables

### ISSUE THREAD

**Issue by davidascher** [participedia/api#85]
Reading 12.2.2 on https://www.postgresql.org/docs/9.5/static/textsearch-tables.html#TEXTSEARCH-TABLES-INDEX makes me feel that we probably want to encode language in the localized text tables using the postgres version of languages, so we can do 

`CREATE INDEX body_idx ON case__localized_texts USING GIN (to_tsvector(language, body));`

but today language values are things like 'en', while postgres knows how to deal with:

```
 pg_catalog | danish     | configuration for danish language
 pg_catalog | dutch      | configuration for dutch language
 pg_catalog | english    | configuration for english language
 pg_catalog | finnish    | configuration for finnish language
 pg_catalog | french     | configuration for french language
 pg_catalog | german     | configuration for german language
 pg_catalog | hungarian  | configuration for hungarian language
 pg_catalog | italian    | configuration for italian language
 pg_catalog | norwegian  | configuration for norwegian language
 pg_catalog | portuguese | configuration for portuguese language
 pg_catalog | romanian   | configuration for romanian language
 pg_catalog | russian    | configuration for russian language
 pg_catalog | simple     | simple configuration
 pg_catalog | spanish    | configuration for spanish language
 pg_catalog | swedish    | configuration for swedish language
 pg_catalog | turkish    | configuration for turkish language
```

**Comment by davidascher** [comment]
As to why we need indexing: I'm getting a time to first byte of 6.5seconds on a simple word search for 'budgeting'.

**Comment by davidascher** [comment]
FYI: https://www.compose.com/articles/indexing-for-full-text-search-in-postgresql/

**Comment by dethe** [comment]
Yes, turns out I was overly optimistic when I thought search was OK because tests ran quickly. We didn't have a test for full-text search! Adding that, and for multi-word searches so I can see the problem. Will implement materialized views for searching, and index those. Figuring out where to put the trigger to re-index them, and also which subset of fields to apply full-text search on. Currently searching across nearly every possible text field, but that may not be reasonable. What about:

* [Language] title
* [Language] body
* tags

Anything else?

**Comment by davidascher** [comment]
I think that would be fine to start. Maybe add a query of user names as well.

**Comment by dethe** [comment]
First cut at the materialized views for indexes (and returning subsetted values) is almost ready. I got to learn how to use UNIONs to build joins of (id, title, etc.) across data types so we can return disparate things in search results. Remaining bit is to figure out how to update the materialized views periodically. I'm going to try for every five minutes because it's kind of an expensive operation. That will mean if someone adds a new record (or updates one) the changes won't make it into the full text search for a few minutes.

Getting the full multi-language support will follow after I get the base indexing done.

**Comment by davidascher** [comment]
Very cool.

Can we trigger reindexes using a middleware on POST calls?

**Comment by dethe** [comment]
Don't even need middleware. If we want to trigger on every post we can do it as part of the SQL call to create or modify the data. I worry that it may make submits take too long, but there may be ways around that. It's one of the options.

**Comment by davidascher** [comment]
We could do it after we return to the browser post insert. Anyways, tbd.

**Comment by dethe** [comment]
Fixed in PR #83, modulo some bits that have their own issues.
